### PR TITLE
#643 Actually deprecate all the stuff we've been describing as discouraged or obsolete or even deprecated

### PR DESCRIPTION
**Pull request by rmjarvis**
I made a `GalSimDeprecationWarning` class that gets raised in a `warning.warn` call whenever someone uses a deprecated function or parameter.  It tells the user what function to use instead of what they did use, along with which version of GalSim this feature started being deprecated.  I think I got all the versions correct.  Most were version 1.1.  

Along the way, I noticed a few things that we forgot to do in previous edits.
- Position and GSParams had not been made immutable as we had done for most other classes (pretty much everything except Image). So I did that here.
- We had officially discouraged `setRNG()` for Noise types.  But there were times when we wanted to make a new noise type with a different `rng`.  So I made the `copy` function optionally take an `rng` parameter to use for the new copy of the Noise object.
- The unit tests still used the old (now deprecated) syntax.  This was mostly intentional to make sure that legacy code would still work correctly.  But bad habits leaked in and some newer unit tests also used the old syntax.  Anyway, I switched everything to use the new syntax now.

This is a big PR in terms of number of lines, but there's not actually a lot to look at in detail.  So if possible, I'd like to get a quick turnaround on this one.  I have another issue almost ready to submit that I'd like to get into v1.3 as well, but it was branched off of this branch, so it will be cleaner to wait until this one is merged before doing the PR on it.

### CODE REVIEW

**Comment by rmandelb** [comment]
I should be able to take a look over the weekend or early next week, so hopefully I won't hold things up too long.  (I have to focus on non-GalSim things today.)

File: galsim/deprecated/__init__.py [hunk 1]
```diff
@@ -15,3 +15,29 @@
...
+#       for python system deprecations.  I think GalSim deprecations are more appropriately
+#       considered UserWarnings, which are not ignored by default.
+class GalSimDeprecationWarning(UserWarning):
+    def __init__(self, s):
+        super(GalSimDeprecationWarning, self).__init__(self, s)
+
+def depr(f, v, s1, s2=None):
+    import warnings
+    s = str(f)+' has been deprecated since GalSim version '+str(v)+'.'
+    if s1:
```
Comment (rmandelb) [inline galsim/deprecated/__init__.py]: Isn't s1 required?  Or is the idea that if there is no suggestion to make for the user, it'll be False?

Comment (rmandelb) [inline galsim/deprecated/__init__.py]: Oh I see, you use an empty string for functionality that has no replacement.  Never mind.

File: galsim/deprecated/shear.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions, and the disclaimer given in the documentation
+#    and/or other materials provided with the distribution.
+#
+
+import galsim
+from galsim.deprecated import depr
+
+def Shear_setE1E2(self, e1=0.0, e2=0.0):
+    """Deprecated method that will be deprecated eventually."""
```
Comment (rmandelb) [inline galsim/deprecated/shear.py]: Deprecated method that will be deprecated eventually?  That sounds kind of weird and redundant to me.  How about "Deprecated method to set the Shear to some (e1, e2)"?  (Same comment applies to the other things in shear.py)

File: galsim/noise.py [hunk 3]
```diff
@@ -204,13 +189,18 @@ def GaussianNoise_applyTo(self, image):
...
-Discouraged; will be deprecated.
-"""
 
-def GaussianNoise_copy(self):
-    return _galsim.GaussianNoise(self.getRNG(),self.getSigma())
+def GaussianNoise_copy(self, rng=None):
+    """Returns a copy of the Gaussian noise model.
+
+    By default, the copy will share the BaseDeviate random number generator with the parent
+    instance.  However, you can provide a new rng to use in the copy if want with
```
Comment (rmandelb) [inline galsim/noise.py]: "if want" -> "if you want"

Comment (rmandelb) [inline galsim/noise.py]: (same comment applies below)

File: galsim/deprecated/__init__.py (hunk 1 above)
Comment (rmjarvis) [inline galsim/deprecated/__init__.py]: Right.  I did that for calculateCovarianceMatrix, which we are just removing, not replacing.  I should add a commend to that effect here so developers will know how to call this in that kind of case.

File: galsim/deprecated/shear.py (hunk 2 above)
Comment (rmjarvis) [inline galsim/deprecated/shear.py]: Yes.  I think it used to say something like "Obsolete method that will be deprecated."  I replace obsolete, but forgot to check that the rest of the string was sensible.  I'll change these.

**Comment by rmandelb** [comment]
I looked at the stuff in galsim/ and its subdirectories carefully, and it seemed fine... but I have to admit my eyes glazed over once I was partway through the stuff in tests/.  I suspect that mistakes in that code would simply result in failed tests, so you would've caught them already.

**Comment by rmjarvis** [comment]
> I have to admit my eyes glazed over once I was partway through the stuff in tests/.

lol.  Thanks, Rachel.  I started by doing all the things in galsim/deprecated, which pulled up a huge number of warnings (not failures) in the unit tests.  Then I edited those until all the warnings went away.  So I think I got them all.  I suppose I could have missed one or two within a `if __name__ == '__main__'` block, but if so, someone will notice eventually and fix it then.

**Comment by rmandelb** [comment]
Yes, sounds good to me.

**Comment by rmjarvis** [comment]
I addressed all of Rachel's comments.  Did anyone else want to take a look?  I'll plan to merge Wednesday if I don't hear otherwise before then.

**Pull request merged**